ugc/storage.cpp

```diff
@@ -1,33 +1,434 @@
+    FromJSONObject(node, kTypeKey, type);
+    index.m_type = type;
+
+    uint64_t offset;
+    FromJSONObject(node, kOffsetKey, offset);
+    index.m_offset = offset;
+
+    bool isDeleted;
+    FromJSONObject(node, kIsDeletedKey, isDeleted);
+    index.m_isDeleted = isDeleted;
+    if (isDeleted)
+      numberOfDeleted++;
+
+    bool isSynchronized;
+    FromJSONObject(node, kIsSynchronizedKey, isSynchronized);
+    index.m_isSynchronized = isDeleted;
+
+    string mwmName;
+    FromJSONObject(node, kMwmNameKey, mwmName);
+    index.m_mwmName = mwmName;
+
+    string mwmVersion;
+    FromJSONObject(node, kDataVersionKey, mwmVersion);
+    index.m_dataVersion = mwmName;
+
+    uint32_t featureId;
+    FromJSONObject(node, kFeatureIdKey, featureId);
+    index.m_featureId = featureId;
+
+    res.emplace_back(move(index));
+  }
+}
+
+string SerializeUGCIndex(vector<Storage::UGCIndex> const & indexes)
+{
+  if (indexes.empty())
+    return string();
+
+  auto array = my::NewJSONArray();
+  for (auto const & i : indexes)
+  {
+    auto node = my::NewJSONObject();
+    auto const & mercator = i.m_mercator;
+    ToJSONObject(*node, kXKey, mercator.x);
+    ToJSONObject(*node, kYKey, mercator.y);
+    ToJSONObject(*node, kTypeKey, i.m_type);
+    ToJSONObject(*node, kOffsetKey, i.m_offset);
+    ToJSONObject(*node, kIsDeletedKey, i.m_isDeleted);
+    ToJSONObject(*node, kIsSynchronizedKey, i.m_isSynchronized);
+    ToJSONObject(*node, kMwmNameKey, i.m_mwmName);
+    ToJSONObject(*node, kDataVersionKey, i.m_dataVersion);
+    ToJSONObject(*node, kFeatureIdKey, i.m_featureId);
+    json_array_append_new(array.get(), node.release());
+  }
+
+  unique_ptr<char, JSONFreeDeleter> buffer(json_dumps(array.get(), JSON_COMPACT | JSON_ENSURE_ASCII));
+  return string(buffer.get());
+}
+
+void SerializeUGCUpdate(json_t * node, UGCUpdate const & update, Storage::UGCIndex const & index)
+{
+  vector<char> data;
+  using Sink = MemWriter<vector<char>>;
+  Sink sink(data);
+  SerializerJson<Sink> ser(sink);
+  ser(update);
+
+  my::Json serializedUgc(data.data());
+  node = serializedUgc.get();
+  auto embeddedNode = my::NewJSONObject();
+  ToJSONObject(*embeddedNode, kDataVersionKey, index.m_dataVersion);
+  ToJSONObject(*embeddedNode, kMwmNameKey, index.m_mwmName);
+  ToJSONObject(*embeddedNode, kFeatureIdKey, index.m_featureId);
+  ToJSONObject(*node, kFeatureKey, *embeddedNode);
+  embeddedNode.release();
+}
+}  // namespace
+
+Storage::Storage(Index const & index)
+  : m_index(index)
 {
   Load();
 }
 
-UGCUpdate const * Storage::GetUGCUpdate(FeatureID const & id) const
+UGCUpdate Storage::GetUGCUpdate(FeatureID const & id) const
 {
-  auto const it = m_ugc.find(id);
-  if (it != end(m_ugc))
-    return &it->second;
-  return nullptr;
+  ASSERT_THREAD_CHECKER(m_threadChecker, ());
+  if (m_UGCIndexes.empty())
+    return {};
+
+  auto const feature = GetOriginalFeature(id);
+  CHECK(feature, ());
+  auto const & mercator = feature->GetCenter();
```

Comment:
GetCenter works for POI's only. Either add a CHECK that feature is a POI, or use feature::GetCenter.